internal/gps/prune.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// PruneOptions represents the pruning options used to write the dependecy tree.
+type PruneOptions uint8
+
+const (
+	// PruneNestedVendorDirs indicates if nested vendor directories should be pruned.
+	PruneNestedVendorDirs = 1 << iota
+	// PruneUnusedPackages indicates if unused Go packages should be pruned.
+	PruneUnusedPackages
+	// PruneNonGoFiles indicates if non-Go files should be pruned.
+	// LICENSE & COPYING files are kept for convience.
+	PruneNonGoFiles
+	// PruneGoTestFiles indicates if Go test files should be pruned.
+	PruneGoTestFiles
+)
+
+var (
+	preservedNonGoFiles = []string{
+		"LICENSE",
+		"COPYING",
+	}
+)
+
+func Prune(baseDir string, l Lock, options PruneOptions, logger *log.Logger) error {
+	if (options & PruneNestedVendorDirs) != 0 {
+		if err := pruneNestedVendorDirs(baseDir); err != nil {
+			return err
+		}
+	}
+
+	if (options & PruneUnusedPackages) != 0 {
+		if l == nil {
+			return errors.New("pruning unused packages requires passing non-nil Lock")
+		}
+		if err := pruneUnusedPackages(baseDir, l, logger); err != nil {
+			return err
+		}
+	}
+
+	// if (options & PruneNonGoFiles) != 0 {
```

Comment:
Yes. 😓 

I actually completed the PR but did a `git checkout -- .` by mistake and now I have to do it again. 😫 

I'm hoping that I'll be able to complete the remaining changes, again, tomorrow.